### ISSUE TITLE
TopoLMsimple.py

### ISSUE THREAD

**Issue by shandingwangyue** [shandingwangyue/TopoLM#3]
"""
TopoLM: Topological Domain Attention - Proof of Concept
-------------------------------------------------------
Implementation of Spatial-Indexed Attention based on 
the 9-Level Token Stratification architecture.
"""

import torch
import torch.nn as nn
import torch.nn.functional as F

class TopologicalDomainAttention(nn.Module):
    def __init__(self, d_model, num_heads):
        super().__init__()
        self.d_model = d_model
        self.num_heads = num_heads
        self.qkv_proj = nn.Linear(d_model, 3 * d_model)
        self.out_proj = nn.Linear(d_model, d_model)

    def forward(self, x, domain_mu, domain_sigma, threshold=0.05):
        """
        Args:
            x: Input token vectors [batch_size, seq_len, d_model]
            domain_mu: Semantic center of the active Macro-Domain [d_model]
            domain_sigma: Variance of the active Macro-Domain [d_model]
            threshold: Probability density threshold for topological inclusion
        """
        batch_size, seq_len, _ = x.size()

        # Step 1: Calculate Topological Inclusion (Probability Density)
        diff = x - domain_mu
        mahalanobis_sq = (diff ** 2) / (domain_sigma + 1e-9)
        inclusion_prob = torch.exp(-0.5 * mahalanobis_sq.mean(dim=-1)) 

        # Step 2: Generate Spatial Mask
        domain_mask = (inclusion_prob > threshold).float()
        attn_mask = domain_mask.unsqueeze(1).unsqueeze(2) 

        # Step 3: Standard QKV Projection
        qkv = self.qkv_proj(x).chunk(3, dim=-1)
        q, k, v = map(lambda t: t.view(batch_size, seq_len, self.num_heads, -1).transpose(1, 2), qkv)

        # Step 4: Spatial-Indexed Attention Calculation
        scores = torch.matmul(q, k.transpose(-2, -1)) / (q.size(-1) ** 0.5)
        scores = scores.masked_fill(attn_mask == 0, float('-inf'))

        attn_weights = F.softmax(scores, dim=-1)
        out = torch.matmul(attn_weights, v)

        out = out.transpose(1, 2).contiguous().view(batch_size, seq_len, self.d_model)
        return self.out_proj(out), inclusion_prob

if __name__ == "__main__":
    torch.manual_seed(42)
    d_model = 256
    
    topo_attn = TopologicalDomainAttention(d_model, num_heads=8)
    
    # Define Immutable Macro-Domain Prior: D_physics
    mu_physics = torch.zeros(d_model)
    sigma_physics = torch.ones(d_model) * 0.5 
    
    # Input sequence: ["engine", "efficiency", "magic_qi", "perpetual_motion"]
    tokens = torch.cat([
        torch.randn(1, 2, d_model) * 0.4,           # In-domain physical concepts
        torch.randn(1, 2, d_model) * 3.0 + 8.0      # Out-of-domain / impossible concepts
    ], dim=1)
    
    output, probs = topo_attn(tokens, mu_physics, sigma_physics, threshold=0.01)
    
    print("--- Topological Inclusion Probabilities & Masking ---")
    token_names = ["engine", "efficiency", "magic_qi", "perpetual_motion"]
    for i, p in enumerate(probs[0]):
        status = "ACCEPTED" if p > 0.01 else "REJECTED (Masked)"
        print(f"Token {i} [{token_names[i]:<16}]: Prob = {p.item():.4f} -> {status}")
